### ISSUE TITLE
Use Caffeine instead of homegrown cache for CustomResourceCache

### ISSUE THREAD

**Issue by metacosm** [java-operator-sdk/java-operator-sdk#356]
We need to provide a cache that can properly support eviction and we should avoid writing one ourselves. [Caffeine](https://github.com/ben-manes/caffeine) looks like a good candidate for that. /cc @csviri

**Comment by csviri** [comment]
When I was implementing that , was thinking about caffein, but here we don't need automatic cache eviction and featured that caffein provides. In a normal service where you would limit the number of objects cached, and have a nice configurable way of doing caching caffein is the weapon of choice. 
However here we actually want to always store the latest CR, so a concurrent hash map was a KISS solution. Although as related to the other issue, we could do this in a configurable way, so open also for other caching ligraries like coffein.

**Comment by metacosm** [comment]
The problem is that the cache will keep growing indefinitely as it is. Imagine a scenario where an operator provisions lots of of long-lived resources which are not modified afterwards. In that scenario, the cache will grow to be quite big but contain resources that are not needed anymore.

The problem is that `CustomResourceCache` is called cache but it really isn't because the `EventHandler` actually doesn't work if there's a "cache" miss so this architecture needs to be cleaned up, imo.

**Comment by csviri** [comment]
@metacosm yes, that is intentional. So that what I mean when we want to have automatic removal of object from cache there are other parts to change, so its not that trivial. 

Although in practice this is not really a problem with not extreme amunt of CR's (thus there are no millions of CRs). Note also think about if that in case an operator restarts, we also reconcile all the objects to they will and up all in the cache at least for a certain amount of time. So there is a point where we anyway cache all the object even with caffein. 

So the point is that this is much more complex problem then just replacing the cache implementation. And several parts are effected. We have to think also about this reconciliation strategy on startup I think, currently that could be also a self inflected DoS attack if there is a huge amount of CR.

**Comment by csviri** [comment]
What we agreed on:
1. CustomResourceEvent source can cache the resources, and transparently get them from the server if there is a miss
2. The cache will be pluggalble, so anyone can specify a cache implementation (we will have a Cache interface) and replace it. And the default can be caffein with some reasonable defaults. (Evicting oldest object first)

**Comment by lburgazzoli** [comment]
@metacosm @csviri on this topic, should the operator sdk leverage the informer provided by the fabric8 client ?

**Comment by csviri** [comment]
@lburgazzoli yes it should, when we did the current implementation thge informers were not there yet.  Started to working on that with other improvements, but some changes will require quite major changes. Did not have the time.

**Comment by metacosm** [comment]
Looking at leveraging informers instead of homegrown cache is something I plan to explore soon.

**Comment by csviri** [comment]
One another thing to think about what would be the nices way to share the caches between event source, somthing like:
https://pkg.go.dev/k8s.io/client-go@v0.22.1/tools/cache#SharedIndexInformer

**Comment by csviri** [comment]
Looking on related issue: https://github.com/java-operator-sdk/java-operator-sdk/issues/522
We might want to do this in 2 steps:
1. Migrate CustomResourceEventSource to Informers (this should be failry quicky & easy )
2. Implenent improved caching (this might take longer)

**Comment by csviri** [comment]
For now it seems that we will just migrating to informers would be good enough. Having a cache with some smart eviction functionality would mean that we cannot use Informers as our backbone for `CustomResourceEventSource`. Which would mean basically reimplementing Informers in a different way.
